File: python/백준/그래프/다이젝스트라/1916 최소비용 구하기.py
import sys
import heapq

input = sys.stdin.readline
INF = int(1e9)

# 도시 수 (노드), 버스 수 (간선)
N = int(input().strip())
M = int(input().strip())

# 그래프 초기화 (1-indexed)
graph = [[] for _ in range(N + 1)]
distance = [INF] * (N + 1)

# 버스 정보 입력
for _ in range(M):
    a, b, cost = map(int, input().split())
    graph[a].append((b, cost))  # 단방향 그래프

# 다익스트라 함수
def dijkstra(start):
    distance[start] = 0
    q = [(0, start)]  # (비용, 노드)

    while q:
        dist, now = heapq.heappop(q)
        if distance[now] < dist:
            continue  # 이미 방문한 노드면 스킵

        for next_node, next_cost in graph[now]:
            cost = dist + next_cost
            if cost < distance[next_node]:  # 더 짧은 경로 발견 시 갱신
                distance[next_node] = cost
                heapq.heappush(q, (cost, next_node))

# 구간 정보 입력 (출발 도시, 도착 도시)
start, end = map(int, input().split())

# 최단 거리 계산 및 출력
dijkstra(start)
print(distance[end])

File: cpp/백준/그래프/다이젝스트라/1916 최소비용 구하기.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <utility>
using namespace std;

const int INF = 1e9;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M;
    cin >> N;
    cin >> M;

    vector<vector<pair<int, int>>> graph(N + 1);
    vector<int> distance(N + 1, INF);

    for (int i = 0; i < M; ++i) {
        int a, b, cost;
        cin >> a >> b >> cost;
        graph[a].push_back({b, cost});
    }

    int start, endNode;
    cin >> start >> endNode;

    distance[start] = 0;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> q;
    q.push({0, start});

    while (!q.empty()) {
        auto [dist, now] = q.top();
        q.pop();

        if (distance[now] < dist) continue;

        for (auto [next_node, next_cost] : graph[now]) {
            int cost = dist + next_cost;
            if (cost < distance[next_node]) {
                distance[next_node] = cost;
                q.push({cost, next_node});
            }
        }
    }

    cout << distance[endNode] << "\n";
    return 0;
}
